superdesk/core/resources/resource_rest_endpoints.py

```diff
@@ -563,10 +572,69 @@ def _build_resource_hateoas(self, req: SearchRequest, doc_count: Optional[int],
 
         return links
 
-    def _populate_item_hateoas(self, request: Request, item: dict[str, Any]) -> dict[str, Any]:
+    def _populate_hateoas_if_needed(self, request: Request, items: list[dict[str, Any]]):
+        """
+        Populate the hateoas information for a list of items only if the
+        endpoint config `populate_item_hateoas` flag is set to True
+        """
+
+        if not self.endpoint_config.populate_item_hateoas:
+            return
+
+        for item in items:
+            self._populate_item_hateoas(request, item, force_append_id=True)
+
+    def _populate_item_hateoas(
+        self, request: Request, item: dict[str, Any], force_append_id: bool = False
+    ) -> dict[str, Any]:
         item.setdefault("_links", {})
         item["_links"]["self"] = {
             "title": self.resource_config.title or self.resource_config.data_class.__name__,
-            "href": self.gen_url_for_item(request, item["_id"]),
+            "href": self.gen_url_for_item(request, item["_id"], force_append_id),
         }
+
+        # extract related links from the `resource_config.data_class` relations (annotations)
+        related_links = self._get_related_links_from_annotations(item)
+        if related_links:
+            item["_links"]["related"] = related_links
+
         return item
+
+    def _get_related_links_from_annotations(self, item: dict[str, Any]) -> dict[str, Any]:
+        """Extract related links from the `resource_config.data_class` annotations.
+
+        Examines the data class annotations for Annotated fields that contain AsyncValidator metadata.
+        For each field with a resource name specified in its validator, generates a related link
+        if the field has a value in the item.
+        """
+        related_links = {}
+        for field_name, annotation in get_annotations(self.resource_config.data_class).items():
```

Comment:
`get_annotations` does not provide annotations from mixin or super classes (known limitation). I came across this with validations. Instead we have the following:
https://github.com/superdesk/superdesk-core/blob/c39fc7c519d85a96f67adcf2ff18faf7d64cb985/superdesk/core/resources/model.py#L243-L249 

Might be good to put the above code in a util to be used by both places.

Reply:
Also, I fear this might cause a bit of a performance issue. If the response includes 200 items, then we're calculating the related fields for each item.

A quick solution might be to store the related fields metadata on the ResourceRestEndpoints class instance, generate it once (on `__init__` maybe), and use that list thereafter.

Reply:
Good observation. I extracted part of the logic into a separate util function + relocated some of the things into the ResourceModel itself so the related attributes are calculated on class initialization and we do not need to do it over and over again. Please take another look.